### PR TITLE
increase tolerance on helical shapes to avoid crash / broken geometry 

### LINKED ISSUE #9377: [Problem] PartDesign: Helix insufficient precision
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Forums discussion

_No response_

### Version

0.20 (Release)

### Full version info

```shell
[code]
OS: Ubuntu Core 20 (ubuntu:GNOME/ubuntu)
Word size of FreeCAD: 64-bit
Version: 0.20.2.29177 +426 (Git) Snap 517
Build type: Release
Branch: tag: 0.20.2
Hash: 930dd9a76203a3260b1e6256c70c1c3cad8c5cb8
Python 3.8.10, Qt 5.15.5, Coin 4.0.0, Vtk 7.1.1, OCC 7.6.3
Locale: English/United States (en_US)
Installed mods: 
  * fasteners 0.4.56
  * Defeaturing 1.2.0
  * Assembly4 0.12.7
  * Dracula 0.0.4
  * A2plus 0.4.60k
  * Behave-Dark-Colors 0.1.1
[/code]
```

### Subproject(s) affected?

PartDesign

### Issue description

There is some unexpected float precision error with Helix in Part Design Workbench. 
1) Pad(50mm) on gear curve.  Create sketch (shifted 50 mm) and pad it. Works well.

2) Helix (50 mm) on gear curve. Create sketch (shifted 50 mm) and pad it. => Two solids. 
expected => one solid 

Screenshot from 2023-04-24 20-27-40
Screenshot from 2023-04-24 20-28-05

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #11083: [Problem] PartDesign: Rendering issue in Subtractive Helix (OCC related)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

The rendering of a SubtractiveHelix depends subtly on one parameter of the used sketch to be subtracted. The thread shows a "pimple" in the generated thread.
Screenshot_20231018_110830
Also the smoothness in nearby threads looks very deteriorated.
The "real" problem then results from the fact that also for an export (e.g. as *.AMF) for a 3D-print this error is persistent.
When the diameter of the underlying sketch is changed from 23mm to 22.99 (encircled) everything looks O.K.
Screenshot_20231018_110900

### Full version info

```shell
[code]
OS: Arch Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.34667 (Git) AppImage
Build type: Release
Branch: main
Hash: af5b8714a8371162addb97dd25c29250fcecf7ad
Python 3.10.12, Qt 5.15.8, Coin 4.0.0, Vtk 9.2.5, OCC 7.6.3
Locale: English/United States (en_US)
[/code]
```

### Subproject(s) affected?

None

### Anything else?

[bug-suspect_Sketch020_23mm-vs-22.9mm.FCStd.zip](https://github.com/FreeCAD/FreeCAD/files/12999834/bug-suspect_Sketch020_23mm-vs-22.9mm.FCStd.zip)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #17113: PartDesign: Helix: crash - endless recursion when creating helix from sketch around cylinder
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

The following file consists of a cylinder and a simple triangular sketch on it's rim. its perfect to make a helix in PartDesign

if you select the sketch and click the helix tool, Freecad will enter an endless-recursion which ends up in a segfault

I could not get a useful backtrace from that since the recursion at the time of crash is tens of thousands of levels deep

[helixBig.zip](https://github.com/user-attachments/files/17298640/helixBig.zip)

### Full version info

```shell
OS: Ubuntu 20.04.6 LTS (XFCE/xfce)
Word size of FreeCAD: 64-bit
Version: 1.1.0dev.38925 (Git)
Build type: Debug
Branch: main
Hash: 12c6fc2920d083c7407419f8b9ae3651798c005a
Python 3.8.10, Qt 5.12.8, Coin 4.0.0, Vtk 7.1.1, OCC 7.6.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/Qt default
```

### Subproject(s) affected?

PartDesign

### Anything else?

**Contributing factor:**

The part in question has a diameter of **250km** (pad) and a height of **10km**.  

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by AIRCAP**
fix #17113
fix #9377
fix #11083

Helical shapes in PartDesign (and to a lesser degree helical paths) suffer from a precision flaw (see #9337) -- cylinders, cones, spheres, boxes, etc... are all "perfect" shapes in OCCT - ruled by exact mathematical formulas where their edges and surfaces are.

But OCCT has no primitive shape for helical shapes - be it 2d and 3d - so helical paths are approximated with bsplines and helicap solids are pipes along these bsplines.

now the issue is this approximation makes errors far in excess of Precision::Confusion() - so boolean operations between helical shapes and primitives (perfect cylinders and cones especially) will fail, as one is a perfect circular shape and the other is "wonky"

The FixBooleans patch alone can not address this. It just deals with floating point accuracy issues in booleans depending on part size. the error helical paths make is several orders of magnitude larger.

The solution in this PR is to increase the shape tolerances of helical shapes by that much, in order to allow boolean operations with adjacent perfect circular shapes to succeed.  This does not completely make helical shapes unproblematic (they are still wonky due to the bspline approximation) but they won't break all the time and can actually be usable.

This PR depends on both #17119  and #17349 as pre-requirements in order to fix the issues listed above. only the combination of these PRs can fully address all of these issues.

### CODE REVIEW

**Comment by AIRCAP** [comment]
Note: merging this without the other 2 PRs should fix the crash in #17113 - but the shape will still be broken and very very slow to recompute

**Comment by AIRCAP** [comment]
hmm i missed something, should have checked CLI tests...

**Comment by AIRCAP** [comment]
ok, all tests pass now.

Note: For helical shapes in PartDesign I added a new Property "Tolerance" - defaults to 0.1 -- which is multiplicative with Precision::Confusion() and size

overall min Prevision = Tolerance * size * Precision::Confusion()

the default 0.1 is enough for most additive helix to fuse and work in most cases. but for extreme cases like 11083 you need it as high as 1000 to succeed (only for the outer thread)

this is exposed to the user to fix when they have problems with these kind of parts

**Comment by AIRCAP** [comment]
here's some testcases
[helix_testcases.zip](https://github.com/user-attachments/files/17452080/helix_testcases.zip)

**Comment by AIRCAP** [comment]
I added test cases for additive and subtractive helices in PartDesign across different size ranges (10^-1 to 10^8 mm )

**Note**: Some of the tests are expected to fail at the moment, until #17119 has been merged.

**Comment by yorikvanhavre** [comment]
Waiting for https://github.com/FreeCAD/FreeCAD/pull/17119

File: src/Mod/PartDesign/App/FeatureHelix.cpp [hunk 1]
```diff
@@ -237,6 +248,8 @@ App::DocumentObjectExecReturn* Helix::execute()
         if (SC.State() == TopAbs_IN) {
             result.Reverse();
         }
+
```
Comment (FlachyJoe) [inline src/Mod/PartDesign/App/FeatureHelix.cpp:251]: Could you fix the lint warnings? Your IDE should be configured to auto-remove the trailing spaces.

File: src/Mod/PartDesign/PartDesignTests/TestHelix.py [hunk 2]
```diff
@@ -129,6 +129,123 @@ def testRectangle(self):
...
+            self.Doc.recompute()
+
+            helix = self.Doc.addObject("PartDesign::AdditiveHelix","AdditiveHelix")
+            body.addObject(helix)
+            helix.Profile = gearSketch
+            helix.ReferenceAxis = (gearSketch,"V_Axis")
+            helix.Placement = FreeCAD.Placement(FreeCAD.Vector(0,0,0), FreeCAD.Rotation(FreeCAD.Vector(0,0,1),0), FreeCAD.Vector(0,0,0))
+
+            helix.Pitch = 2*(10**exponent)
+            helix.Turns = 3
```
Comment (FlachyJoe) [inline src/Mod/PartDesign/PartDesignTests/TestHelix.py:151]: Turns count is not relevant, you can decrease it to speed up the test.

File: src/Mod/PartDesign/App/FeatureHelix.cpp [hunk 3]
```diff
@@ -229,6 +234,12 @@ App::DocumentObjectExecReturn* Helix::execute()
 
         TopoDS_Shape face = sketchshape;
         face.Move(invObjLoc);
+
+        Bnd_Box bounds;
+        BRepBndLib::Add(path, bounds);
+        double size=sqrt(bounds.SquareExtent());
+        ShapeFix_ShapeTolerance fix;
+        fix.LimitTolerance(path, Precision::Confusion() * 1e-6 * size ); // needed to produce valid Pipe for very big parts
```
Comment (FlachyJoe) [inline src/Mod/PartDesign/App/FeatureHelix.cpp:242]: You introduce a new hard coded parameter, can you explain the choice of the value?

**Comment by AIRCAP** [comment]
#17119 and #17349 have now both been merged into main, so this should now pass all tests (once the CI wakes from its slumber)

backport to 1.0 depends on  #17325 getting merged to 1.0

**Comment by AIRCAP** [comment]
We have a weird CI failure.

Ubuntu 20.04 is a invalid shape created in one of the tests - issue with OCCT 7.3 -- I can look into that.

Windows is a build system dependency / CI issue. Is that python build 32bit?

```
024-10-21T23:33:11.3110304Z OverflowError: Python int too large to convert to C long
2024-10-21T23:33:11.3110834Z 
2024-10-21T23:33:11.3111235Z The above exception was the direct cause of the following exception:
2024-10-21T23:33:11.3111827Z 
2024-10-21T23:33:11.3112024Z Traceback (most recent call last):
2024-10-21T23:33:11.3113004Z   File "C:\FC\build\release\Mod\PartDesign\PartDesignTests\TestHelix.py", line 138, in testGiantHelix
2024-10-21T23:33:11.3114564Z     TestSketcherApp.CreateRectangleSketch(gearSketch, (10*(10**exponent), 0), (1*(10**exponent), 1*(10**exponent)))
2024-10-21T23:33:11.3116305Z   File "C:\FC\build\release\Mod\Sketcher\SketcherTests\TestSketcherSolver.py", line 67, in CreateRectangleSketch
2024-10-21T23:33:11.3118056Z     SketchFeature.addConstraint(Sketcher.Constraint("DistanceX", i + 2, 2, corner[0]))
2024-10-21T23:33:11.3118996Z                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-10-21T23:33:11.3120102Z SystemError: <class 'Sketcher.Constraint'> returned a result with an exception set
```

**Comment by AIRCAP** [comment]
I don't understand that. I have Ubuntu 20.04 and I did compile against OCCT 7.3.0 to test this. Why is the CI still failing - and with invalid shape? I never had an invalid shape failure ?!?

**Comment by yorikvanhavre** [comment]
Still waiting for https://github.com/FreeCAD/FreeCAD/pull/17119 to be tested, issues fixed and backported

**Comment by chennes** [comment]
Noting that the backport of #17119 is #17325.

**Comment by yorikvanhavre** [comment]
Leaving this open for another week to see where the forum discussion is going. Decision to merge or not to be taken next monday

**Comment by yorikvanhavre** [comment]
backported

**Pull request merged**